src/Build/Logging/BinaryLogger/BinaryLogReplayEventSource.cs

```diff
@@ -85,24 +170,134 @@ public void Replay(string sourceFilePath, CancellationToken cancellationToken)
         public void Replay(BinaryReader binaryReader, CancellationToken cancellationToken)
+            => Replay(binaryReader, false, cancellationToken);
+
+        /// <summary>
+        /// Read the provided binary log file and raise corresponding events for each BuildEventArgs
+        /// </summary>
+        /// <param name="binaryReader">The binary log content binary reader - caller is responsible for disposing, unless <paramref name="closeInput"/> is set to true.</param>
+        /// <param name="closeInput">Indicates whether the passed BinaryReader should be closed on disposing.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> indicating the replay should stop as soon as possible.</param>
+        public void Replay(BinaryReader binaryReader, bool closeInput, CancellationToken cancellationToken)
         {
-            int fileFormatVersion = binaryReader.ReadInt32();
+            using var reader = OpenBuildEventsReader(binaryReader, closeInput, AllowForwardCompatibility);
+            NotificationsSourceCreated?.Invoke(reader);
+            Replay(reader, cancellationToken);
+        }
 
-            // the log file is written using a newer version of file format
-            // that we don't know how to read
-            if (fileFormatVersion > BinaryLogger.FileFormatVersion)
+        /// <summary>
+        /// Read the provided binary log file and raise corresponding events for each BuildEventArgs
+        /// </summary>
+        /// <param name="reader">The build events reader - caller is responsible for disposing.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> indicating the replay should stop as soon as possible.</param>
+        public void Replay(BuildEventArgsReader reader, CancellationToken cancellationToken)
+        {
+            bool supportsForwardCompatibility = reader.FileFormatVersion >= 18;
+
+            // Allow any possible deferred subscriptions to be registered
+            if (HasStructuredEventsSubscribers || !supportsForwardCompatibility)
             {
-                var text = ResourceUtilities.FormatResourceStringStripCodeAndKeyword("UnsupportedLogFileFormat", fileFormatVersion, BinaryLogger.FileFormatVersion);
-                throw new NotSupportedException(text);
+                _onStructuredReadingOnly?.Invoke();
+            }
+            else
+            {
+                _onRawReadingPossible?.Invoke();
             }
 
-            using var reader = new BuildEventArgsReader(binaryReader, fileFormatVersion);
-            NotificationsSourceCreated?.Invoke(reader);
+            reader.EmbeddedContentRead += _embeddedContentRead;
+            reader.ArchiveFileEncountered += _archiveFileEncountered;
+            reader.StringReadDone += _stringReadDone;
+            reader.StringEncountered += _stringEncountered;
```

Comment:
nit: Is there any point in undoing this before the method returns? I'm thinking through the hypothetical scenario of the caller calling `Replay`, canceling the operation half-way through using the cancellation token, then calling again to resume. I understand it's not the intended use case but still wondering what exactly would happen.

Reply:
Bit edgy - but still possible and events would then be delivered twice - so good point. Unsubscribe will not hurt - adding it